lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/LabelledCharArrayMatcher.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search.uhighlight;
+
+import org.apache.lucene.util.UnicodeUtil;
+import org.apache.lucene.util.automaton.ByteRunAutomaton;
+
+/**
+ * Associates a label with a CharArrayMatcher to distinguish different sources for terms in highlighting
+ *
+ * @lucene.internal
+ */
+public interface LabelledCharArrayMatcher extends CharArrayMatcher {
```

Comment:
If CharArrayMatcher is now specific to the UnifiedHighlighter, then why have a "LabelledCharArrayMatcher" separate?  Lets just have CharArrayMatcher that has a Label?

Reply:
The advantage of having both is that `CharArrayMatcher` is a function interface, so we can pass `CharacterRunAutomaton::run` as one

Reply:
An extra class just so that we have a functional interface in one spot internally?  I wouldn't make that trade-off but I leave this to you.

Reply:
There's also the fact that not every matcher requires a label - ones created from a set of terms, for example, in the MemoryIndexOffsetStrategy.  We can probably clear this up later, but I'd rather keep things as they are for now and get this in.